Wikipedia/Code/WMFAppViewController+Extensions.swift

```diff
@@ -97,13 +100,96 @@ extension WMFAppViewController: NotificationsCenterPresentationDelegate {
+            self?.navigationController?.pushViewController(notificationsCenterViewController, animated: true)
+        }
+
+        guard editingFlowIsInHierarchy else {
+            dismissAndPushBlock()
+            return
+        }
+        
+        presentEditorAlert(on: topMostViewController, confirmationBlock: dismissAndPushBlock)
+    }
+    
+    private var editingFlowIsInHierarchy: Bool {
+        var currentController: UIViewController? = navigationController
+
+        while let presentedViewController = currentController?.presentedViewController {
+            if let presentedNavigationController = (presentedViewController as? UINavigationController) {
+                for viewController in presentedNavigationController.viewControllers {
+                    if viewController.isPartOfEditorFlow {
+                        return true
+                    }
+                }
+            } else if presentedViewController.isPartOfEditorFlow {
+                return true
+            }
+            
+            currentController = presentedViewController
+        }
+
+        return false
+    }
+    
+    private var topMostViewController: UIViewController? {
+            
+        var topViewController: UIViewController = navigationController ?? self
+
+        while let presentedViewController = topViewController.presentedViewController {
+            topViewController = presentedViewController
+        }
+
+        return topViewController
+    }
+    
+    private func presentEditorAlert(on viewController: UIViewController, confirmationBlock: @escaping () -> Void) {
+        
+        let title = CommonStrings.editorExitConfirmationTitle
+        let message = CommonStrings.editorExitConfirmationBody
+        
+        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let discardAction = UIAlertAction(title: CommonStrings.discardEditsActionTitle, style: .destructive) { _ in
+            confirmationBlock()
+        }
+        let cancelAction = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel)
+        
+        alertController.addAction(discardAction)
+        alertController.addAction(cancelAction)
+        
+        viewController.present(alertController, animated: true, completion: nil)
+        
+    }
+}
+
+fileprivate extension UIViewController {
+    var isPartOfEditorFlow: Bool {
+        return (self is SectionEditorViewController ||
+                self is InsertLinkViewController ||
+                self is InsertMediaViewController ||
+                self is EditPreviewViewController ||
+                self is InsertMediaSearchResultPreviewingViewController ||
+                self is EditSaveViewController ||
+                self is DescriptionEditViewController)
+    }
+    
+    var isDisplayingNotificationsCenter: Bool {
+        
+        if let navigationController = (self as? UINavigationController),
```

Comment:
When the user is in the Notifications Center and has a modal view presented (Filters or Inbox), tapping a notification pushes another Notifications Center on to the stack. It seems like we'd probably want to avoid that. 

A lightweight solution may be to include checking if `self` is a `DisappearingCallbackNavigationController` here, though if you go this route I'd strongly recommend changing the name of `DisappearingCallbackNavigationController` to something prefixed with `NotificationsCenter` or something to make the functionality clearer for anyone in the future who may try and reuse it. 

A stronger, though heavier weight, solution (which also applies to the editing flow above) is probably having a protocol that represents the editing flow and one for the notifications center itself and having certain view controllers just conform to those empty protocols.

I think either of these approaches (or if you think of something else I'm missing) would be fine.

Reply:
Great call - this is fixed in the latest commit.